src/gadgets/scalar/range.rs

```diff
@@ -1,204 +1,216 @@
 //! Collection of range-checking gadgets for Bls12_381 scalars.
-use crate::gadgets::GadgetErrors;
-use anyhow::{Error, Result};
 use dusk_plonk::prelude::*;
 
-/// Builds a complex range-proof (not bounded to a pow_of_two) given a
-/// composer, the max range and the witness.
-///
-/// Checks that `witness <= max_range` returning a boolean `Variable` as a result
-/// where `1 = holds` and `0 = Does not hold`.
-pub fn single_complex_rangeproof_gadget(
+/// An allocated scalar holds the underlying witness assignment for the Prover
+/// and a dummy value for the verifier
+/// XXX: This could possibly be added to the PLONK API
+#[derive(Clone)]
+pub struct AllocatedScalar {
+    var: Variable,
+    scalar: BlsScalar,
+}
+
+impl AllocatedScalar {
+    /// Allocates a BlsScalar into the constraint system as a witness
+    pub(crate) fn allocate(composer: &mut StandardComposer, scalar: BlsScalar) -> AllocatedScalar {
+        let var = composer.add_input(scalar);
+        AllocatedScalar { var, scalar }
+    }
+}
```

Comment:
Can you add another method where you can just input the Scalar and the Variable?

Otherways, when you have the Variable but not the Scalar, you need to re-declare the Variable because there's no method that allows you to get `AllocatedScalar`s from a `Variable-Scalar` pair.

Reply:
I have made the var and scalar variables public for interoperability. Although this should be ideally used with just the allocate method.

Reply:
I agree, but it should be the case if this was part of the Plonk constraint system already.

Since it's not, this has to be more flexible, otherways we have to duplicate variables or stuff like this. Do you agree?

Reply:
Yep sure, I don't think it would make a massive difference, _if_ this was later integrated into PLONK